### DISCUSSION TITLE
Decimals of CKB

### DISCUSSION THREAD

**Discussion by Keith-CY** [nervosnet/godwoken_explorer#600] [Q&A] [answered]
I was always confused with CKB's decimals between layer1 and layer2, so I open this discussion to make it clear.

Now there're 3 places to display the amount of CKB.

1. deposit/withdraw CKB
2. Transfer CKB
3. User's balance in CKB

I thought there were two CKB in layer 2, the one used in transfers between users(called CKB), and the other one used to pay the transaction fee(called gCKB).

The CKB transferred between users has deciamls 8, which is the same as that of CKB in layer 1, while the gCKB used to pay fee has decimals 18.

In this case, when user A deposits 1 CKB(10^8 shannons) from layer 1 to layer 2, 1 CKB(10^8 shannons) will be minted on layer 2, and the record will be displayed as follow,

```
type    |  value(CKB)
deposit |    1
```
and user A's balance is `1 CKB`

then user A transfers 0.5 * 10^8 shannons(0.5 CKB) to others with 100 shannons(100 * 10^-18 pCKB), the record will be displayed as follow

```
type | value(CKB) | fee(pCKB)
send | 0.5        | 100 * 10^-18
```

user A's balance turns into `1 - 0.5 - 100 * 10^-8 = 0.49999999 CKB`

This is my understanding of CKB and pCKB before. 

-----

Today I've learned that there's only one CKB in layer 2, gCKB, which has decimals 18. When 1 CKB(10**8 shannons) in layer 1 is deposited to layer 2, 1 gCKB(10^18 gshannons) will be minted. Then the example mentioned above becomes

```
type    |  value(gCKB)
deposit |  1
```

user A's balance is `1 gCKB`

```
type  |    value(gCKB)         | fee(gCKB)
send  |    0.5 * 10^-10       |  100 * 10^-18 
```

user A's balance is `1 - 0.5 * 10^-10 - 100 * 10^18 = 0.9999999999499999 gCKB`

Is the second case correct? @Flouse @gpBlockchain

**Comment by Keith-CY** [comment]
cc @classicalliu, @zeroqn

**Comment by Keith-CY** [comment, marked as answer]
Confirmed with @gpBlockchain and @classicalliu that the second case is correct

**Reply by Flouse** [reply]
Yes, there's only one CKB in layer 2, called gCKB or pCKB, which decimals is 18.

## Related PRs
godwoken
  - https://github.com/nervosnetwork/godwoken/pull/675

godwoken-scripts
  - https://github.com/nervosnetwork/godwoken-scripts/pull/118

godowken-polyjuice
  - https://github.com/nervosnetwork/godwoken-polyjuice/pull/146
